### ISSUE TITLE
Memory leak in the "no-unsupported-features" rule

### ISSUE THREAD

**Issue by bponomarenko** [vuejs/eslint-plugin-vue#1717]
**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**

- **ESLint version:** 7.32.0
- **eslint-plugin-vue version:** 7.20.0
- **Node version:** 14.16.0
- **Operating System:** macOS Catalina 10.15.7

Some additional details:
- Our project is relatively big – around 3000 `*.vue` files and around 2000 of `*.js` files with `// @vue/component` comment

**Please show your full configuration:**

```json5
{
    "root": true,
    "env": { "es2021": true },
    "plugins": ["vue"],
    "extends": ["plugin:vue/recommended"],
    "rules": {
        "vue/no-unsupported-features": ["error", { "version": "^2.6.14" }]
    }
}
```
*Note this is not full `.eslintrc` file, but just relevant part with the configured rule in question*

**What did you do?**	

*No code here because this issue can be reproduced with both valid and invalid code.*

**What did you expect to happen?**
Memory consumption does not significantly change when `no-unsupported-features` rule is enabled.

**What actually happened?**

*ESLint output is not relevant here, as we need to look into "Activity Monitor" instead.*

With the rule enabled, during the lint process (which we run as `eslint --ext .js,.vue .`) RAM consumption keeps raising and peaks at to 3.5G-4Gb level. When we disable the rule – it doesn't go more than 800-900Mb. I tried to disable other rules, and it seems this is the one which causes the issue.

I tried to debug it myself, and I narrowed it down to [this line](https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/utils/style-variables/index.js#L35-L36):
```ts
/** @type {Map<VElement, StyleVariablesContext>} */
const cache = new Map()
```
This file is used in the [`style-css-vars-injection` syntax spec](https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/rules/syntaxes/style-css-vars-injection.js#L7), which is part of the [`no-unsupported-features` rule](https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/rules/no-unsupported-features.js#L28).
From my understanding, this `cache` is created globally per entire lint process, while it would be populated with values from each of the scanned files. This cache uses `VElement` objects as a key, which makes them "trapped" in this cache for the duration of the entire lint process. As a result Node.js cannot garbage collect those references when scan for individual file is done.

Simple change to `WeakMap` seems to be solving the problem:
```ts
/** @type {WeakMap<VElement, StyleVariablesContext>} */
const cache = new WeakMap()
```

Btw during my experiments, I noticed that [part of that code that reads from the `cache`](https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/utils/style-variables/index.js#L55-L58) is **never executed**. I'm not sure what is the use case for this cache implementation, but for now it seems redundant for me and it could be removed altogether.

**Repository to reproduce this issue**

I'm not sure if you still need it, but I can add it if you confirm.

**Comment by ota-meshi** [comment]
Thank you for finding the cause!

**Comment by bponomarenko** [comment]
@ota-meshi Thanks for the quick fix!

Unfortunately I found another memory leak in the another rule. Will report it later today.

**Comment by bponomarenko** [comment]
@ota-meshi That another rule with memory leak is `no-unused-properties`, but it seems the cause is exactly the same – it also uses `style-variables` utility. So the fix you did already should be enough. Thanks again.

**Comment by ota-meshi** [comment]
Thanks for letting me know!
